Desktop: Fixes #5686: Fixed Tags Order

OmGole opened the pull request:
This fixes the bug #5686; The tag order between left panel and windows tag is different.
![tags order](https://user-images.githubusercontent.com/90026187/153363103-7d2b9bf6-0455-4e2a-b926-96eaf0e26949.jpg)
Whenever tags starting with emoji, #, @ were added then the order for tags in the promptDialog dropdown menu and sidebar TAGS section were different.

The cause of the bug was that different sorting logic was used in the sidebar tags section and setTags.ts (the sort method in this file is used to sort tags in the promptDialog dropdown menu).

To fix this I did the following, removed the old sort method in setTags.ts and added the sort method which was used in the side-menu-shared.js (this sort method is used to sort tags in the sidebar). As the sort methods are same the outputs are also the same. I have also added the unit test.

roman-r-m commented on packages/lib/models/Tag.test.js:
```diff
@@ -132,28 +173,109 @@ describe('models/Tag', function() {
-	}));
+	});
 
+	it('should sort tags in promptDialog', async () => {
+		const folder1 = await Folder.save({ title: 'folder1' });
```
This test does not cover your changes at all.
Sure you copied the logic from `setTags` here but now if someone will accidentally breaks `setTags`, your test will still be passing. Which defeats the whole purpose of it.

OmGole replied:
Then how do I test it? I will have to first export it using global variables, right? I am actually writing my first test here so just asking for feedback?

roman-r-m replied:
Your test needs to call `setTags`

OmGole replied:
I will write a new test accordingly, I just need a little push, is exporting a variable from a file for testing allowed? I don't see another way to test without exporting?

OmGole replied:
But won't that test the entire file. I mean I just want to test the startTags and tagSuggestions from the entire file.

roman-r-m replied:
If you don't want to test the whole file (which would be useful btw) you can extract the parts you want to test to separate methods and test them.

OmGole replied:
Okay I will try to test the entire file. Also, extracting the parts using methods will work on classes but setTags uses functions. I researched about it on stackoverflow and found out that in order to export a variable from a function we will have to use global variables.

OmGole commented:
Update: I am planning to wrap the sorting logic in a function so that it can be used to sort tags anywhere in the codebase. This will help us maintain consistency in the sorting logic everywhere and also maintain cleaner code. Wrapping it in a function will also help me export it to a test file and perform unit tests on it. Also thanks to @roman-r-m for all the constructive feedback!!! 

Here sorting logic while be implemented using localeCompare as stated above by @CalebJohn

laurent22 commented:
It looks like the formatting of the code has been changed by prettier or something. Please revert these cosmetic changes and keep only your own changes.

OmGole commented:
Ok

OmGole commented:
Hey @laurent22 my changes  will consist of more than 50 lines of code should I go ahead and do it or make a post about it on the joplin forum?

OmGole commented:
Update: I have reverted back the changes made by prettier. I have added a static method (sortTags) in Tag.ts to sort tags. This method will help us to maintain consistency in sorting tags anywhere in the codebase on top of that it will also allow us to make changes to the sorting functionality less cumbersome. If we have to change the sorting functionality all we have to do is change the sortTags method. I used the sortTags method in three files i.e setTags.ts (used for promptDialog default values and promptDialog dropdown values), Taglist.tsx (used for showing tags in the bottom left corner of the noteEditor), side-menu-shared.js (used for showing tags on the sidebar). Also I have written test for the sortTags method in Tag.test.js

roman-r-m commented on packages/lib/models/Tag.ts:
```diff
@@ -216,4 +216,27 @@ export default class Tag extends BaseItem {
 		});
 	}
+
+	static sortTags(tags: any) {
```
Looks like you're using this function to compare objects of different types in different places: sometimes it's tags, other times it's `{ value: tag.id, label: tag.title }`
It seems to me it's best to rewrite it to always have args of the same time. Then you can also get rid of `any` and give it proper type.

OmGole replied:
sortTags sorts on the basis of the title of the tag so it will work for `{ value: tag.id, label: tag.title }` and tags as well. I can write an if statement for it in sortTags but even without it the functionality will work.

roman-r-m replied:
The issue with how it's written right now is that if someone wanted to use your function elsewhere, it's not possible to understand by looking at the signature what kind parameters it expects. You have to look at its implementation to find that it should be an array of things that have either title or label.

At the same time, you can make a trivial change to specify that this function requires an array of Tags and it makes things much more clear.

laurent22 replied:
For the type, it's possible to use TagEntity. Also I think this function shouldn't support sorting by label. Instead the view should be changed so instead of having Load tags => convert to list items => sort. You should have Load tags => sort => convert to list items. That way you don't need any special case in this function

OmGole commented:
Update: I have added type to tags in `sortTags` as requested and also removed the if statement for label as suggested by laurent

roman-r-m commented on packages/lib/models/Tag.ts:
```diff
@@ -216,4 +216,20 @@ export default class Tag extends BaseItem {
 	}
+
+	static sortTags(tags: TagEntity[]) {
+		if (tags[0].title) {
```
I don't think you need the condition anymore

OmGole replied:
oops!! will remove it asap

laurent22 commented on packages/lib/models/Tag.ts:
```diff
@@ -216,4 +216,18 @@ export default class Tag extends BaseItem {
 	}
+
+	static sortTags(tags: TagEntity[]) {
+		return tags.sort((a: any,b: any)=>{
```
still need types here

laurent22 commented on packages/lib/models/Tag.test.js:
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
+		// sortedTagsTitles contains the titles of all the sorted tags
+		const sortedTagsTitles = Tag.sortTags(defaultTags).map(a => a.label);
+
+		expect(sortedTagsTitles).toEqual(['@⏲15 min', '@⏲30 min', '@⏲60 min', '#coding', '#house', '#wait']);
+		expect(sortedTagsTitles).not.toEqual(['@⏲15 min', '#house','#coding', '@⏲60 min', '#wait', '@⏲30 min']);
```
But the logic... If a = 1, do you think you need to test that a != 2?

laurent22 commented on packages/lib/models/Tag.test.js:
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
+
+		// This test checks the sorting functionality for tags in all three files(setTags, Sidebar and TagList)
+		// commonTags contains unsortedTags
+		const commonTags = await Tag.commonTagsByNoteIds([note1.id]);
```
That doesn't do what you say it's doing. With the way it's implemented now, you don't need to test setTags or the SideBar, so please remove this test.

laurent22 commented on packages/lib/models/Tag.test.js:
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
+		});
+
+		// sortedTagsTitles contains the titles of all the sorted tags
+		const sortedTagsTitles = Tag.sortTags(defaultTags).map(a => a.label);
```
There shouldn't be anything about labels here.

laurent22 commented on packages/lib/models/Tag.test.js:
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
 	}));
 
+	it('should sort tags for setTags, Sidebar and TagList', (async () => {
+		const folder1 = await Folder.save({ title: 'folder1' });
+		const note1 = await Note.save({ title: 'ma note 0', parent_id: folder1.id });
+
+		// original/sorted order
+		const tag1 = await Tag.save({ title: '#coding' });
+		const tag2 = await Tag.save({ title: '#house' });
+		const tag3 = await Tag.save({ title: '#wait' });
+		const tag4 = await Tag.save({ title: '@⏲15 min' });
+		const tag5 = await Tag.save({ title: '@⏲30 min' });
+		const tag6 = await Tag.save({ title: '@⏲60 min' });
+
+		await Tag.addNote(tag4.id, note1.id);
+		await Tag.addNote(tag2.id, note1.id);
+		await Tag.addNote(tag1.id, note1.id);
+		await Tag.addNote(tag6.id, note1.id);
+		await Tag.addNote(tag3.id, note1.id);
+		await Tag.addNote(tag5.id, note1.id);
```
You don't need any of this. Just mock the data - `[{ title:'#coding' }, { title:'#wait' }, etc.`

laurent22 commented on packages/lib/models/Tag.test.js:
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
+
+		expect(sortedTagsTitles).toEqual(['@⏲15 min', '@⏲30 min', '@⏲60 min', '#coding', '#house', '#wait']);
+		expect(sortedTagsTitles).not.toEqual(['@⏲15 min', '#house','#coding', '@⏲60 min', '#wait', '@⏲30 min']);
+	}));
```
Please add more tests - empty list, tags with no title property, etc. Read the code you've copied and pasted in sortTags and make sure you test all cases.

laurent22 commented:
Please don't mark comments as "resolved" because then we don't know what changes have been reviewed or not.

OmGole commented:
Hey @laurent22, I have unresolved the previous conversations can you please review it again?

laurent22 commented:
All good now, thanks @OmGole for adding the tests and making these additional changes.
